### PR TITLE
New attached windows need to be placed

### PR DESCRIPTION
**Pull request by AlanGriffiths**
Alternative to #3675

### CODE REVIEW

**Review by mattkae — CHANGES_REQUESTED** [review]
This does not fix the problem that I was having with `cosmic-bg` because `LayerSurfaceV1::configure()` happens before `BasicWindowManager::place_new_surface()`. The surface has not yet been placed when we send the configure event

**Comment by AlanGriffiths** [comment]
That sequence seems ... wrong. But I see it too:
```
[ 635173.015]  -> wl_compositor#6.create_surface(new id wl_surface#11)
[ 635173.030]  -> zwlr_layer_shell_v1#8.get_layer_surface(new id zwlr_layer_surface_v1#12, wl_surface#11, wl_output#3, 0, "wallpaper")
[ 635173.045]  -> zwlr_layer_surface_v1#12.set_anchor(15)
[ 635173.052]  -> zwlr_layer_surface_v1#12.set_exclusive_zone(-1)
[ 635173.058]  -> zwlr_layer_surface_v1#12.set_keyboard_interactivity(0)
[ 635173.066]  -> wl_surface#11.commit()
[ 635173.073]  -> wp_viewporter#9.get_viewport(new id wp_viewport#13, wl_surface#11)
[ 635173.102]  -> wp_fractional_scale_manager_v1#10.get_fractional_scale(new id wp_fractional_scale_v1#14, wl_surface#11)
[ 635173.788] zwlr_layer_surface_v1#12.configure(9, 0, 0)
[ 635173.807] zwlr_layer_surface_v1#12.configure(10, 1600, 900)
```

**Comment by AlanGriffiths** [comment]
> The surface has not yet been placed when we send the configure event

Fixed now

**Comment by mattkae** [comment]
This makes it not crash :+1:  But it doesn't make it work for `cosmic-bg` unlike my other PR. I'll investigate and let you know why. They appear the same on the surface

**Comment by mattkae** [comment]
I have fixed it with the latest commit @AlanGriffiths:

image

**Comment by Saviq** [comment]
> 32 - wlcs.Anchor/LayerSurfaceLayoutTest.* (Failed)

:)

File: src/server/frontend_wayland/layer_shell_v1.cpp [hunk 1]
```diff
@@ -606,9 +616,13 @@ void mf::LayerSurfaceV1::ack_configure(uint32_t serial)
         return;
     }
 
-    client_size.set_pending(geom::Size{
-        acked_configure_size.width.value_or(client_size.pending().width),
-        acked_configure_size.height.value_or(client_size.pending().height)});
+    auto const& width = acked_configure_size.width
+        ? acked_configure_size.width
+        : client_size.pending().width;
```
Comment (AlanGriffiths) [inline src/server/frontend_wayland/layer_shell_v1.cpp:621]: Isn't `value_or()` clearer?

Comment (AlanGriffiths) [reply src/server/frontend_wayland/layer_shell_v1.cpp:621]: Ignore me. The result is a `std::optional`

**Comment by AlanGriffiths** [comment]
Oops! Pulled the rug out of another corner

**Comment by AlanGriffiths** [comment]
@mattkae I'm not sure what the essential behaviour you want from https://github.com/canonical/mir/pull/3676/commits/9247cc95c8c4a981dc1a43ac2b3ccb34653511d8 is. I've fixed the stuff it broke, but now not sure if the last two commits are even changing anything. (A shame you couldn't add a test)

**Comment by AlanGriffiths** [comment]
Not sure why (but I never tracked down the cause) this also seems to fix some intermittent yambar crashing

**Review by mattkae — APPROVED** [review]
Working well for me now :+1:

**Pull request merged**